### DISCUSSION TITLE
No energy evolution in Grand Potential Phase Field model

### DISCUSSION THREAD

**Discussion by scottedmuller-pnnl** [idaholab/moose#17437] [Q&A Modules: Phase field] [answered]
Hi All,

I am developing a Grand Potential Phase Field model of a binary system in moose and it is having some really odd issues. The energies look right in both phases but I can supersaturate one phase a ton without causing any evolution. I can look at the free energy and see that it is increased appropriately in the given phase, indicating there should be a driving force for the energy to minimize, but no changes occur. Does anyone have any specific advice for troubleshooting a lack of energy minimization in my simulations? The energy is not increasing either there is just no change at all.

Thanks,
Scott

**Comment by laagesen** [comment, marked as answer]
Hi Scott- when you say the total energy is not decreasing, how are you determining that? Are you using a postprocessor to sum the local free energy density? Bear in mind that when you are using a grand potential model, it is the total grand potential that should decrease and not the Helmholtz free energy (although I think the first should imply the second). Also, what are your initial conditions? Just want to make sure it is not stuck in a metastable state.

**Reply by scottedmuller-pnnl** [reply]
Hi Larry,

I output both the grand potential and "regular" free energy as a Material block (just for looking at in Paraview). I've summed over the domain of the Grand Potential free energy density, and it is also not changing, neither increasing or decreasing (unless I change the mobilities in dramatic non-physical ways). The boundary conditions are periodic, and I'm currently considering a 1-D case for simplicity, with a domain of 10 microns. For the initial conditions, I use BoundingBoxIC to make a 4 micron region in the middle of the domain the solid (eta = 1) and everywhere else liquid (eta = 0), and I set the interface width (int_width) to 0.5 microns. 
![pic1](https://user-images.githubusercontent.com/56096041/113194900-8009e500-9216-11eb-90c8-031821a5b5c9.png)
I use a similar BoundingBoxIC for the chemical potential, with the solid region being set to zero (mu = 0), and the liquid region chosen based on the desired level of saturation.
![pic2](https://user-images.githubusercontent.com/56096041/113194901-80a27b80-9216-11eb-8106-c4ef5a22ace1.png)
I've tested a variety of mu values for the liquid, including absurdly high values corresponding to concentrations close to the solid, but it makes no difference in the evolution of the system.

Diffusion in the liquid and solid is on the order of 10^-3 and 10^-5 um^2/us respectively, and chi (partial of omega wrt mu) is on the order of 10^-5. When the mobility of the order parameter is set close to (but still above) the mobility of the chemical potential, nothing evolves at all.

Not sure how helpful this is, but when I artificially increase the diffusion by a few orders of magnitude, and increase the order parameter mobility (eta) to make sure evolution is diffusion-limited, it produces odd behavior in the order parameter after a large number of steps:
![image](https://user-images.githubusercontent.com/56096041/113225500-a80f3d80-9242-11eb-8d5c-dd5ec70744d1.png)
A state that I would think would be impossible. The chemical potential in this case is
![image](https://user-images.githubusercontent.com/56096041/113225635-f7ee0480-9242-11eb-9656-cddf14839fba.png)
Again, not sure how useful this is. I know there are a lot of random things that can be done to mess with a PF model to produce weird non-physical effects, but maybe it can help with troubleshooting?

Please let me know what other information I can give you to help diagnose the problem. Thank you very much for your help!

-Scott

**Reply by laagesen** [reply]
Hi Scott, a couple of questions for you:
-Are you starting from the example file moose/modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialPFM.i? If so please make sure your MOOSE repo is up to date; there was a bug fix to this implemented a few months ago (see https://github.com/idaholab/moose/pull/16343)
-What are the free energies you are using for the solid and liquid phases?
-If you set diffusivities in both phases to 1 does the problem persist?
-If so can you please paste in the solver output from the first time step?

**Reply by scottedmuller-pnnl** [reply]
I developed my model by hand, based mostly on Plapp's formulation from his 2011 paper (http://dx.doi.org/10.1103/PhysRevE.84.031601). I'm using a parabolic free energy for the solid, and I was originally using a free energy derived from the ideal solution model for the liquid, but I interpolated it as a parabolic free energy during troubleshooting, in case there were compatibility issues between the free energies.

I just tested setting the diffusivities in both phases equal to 1, and the problem actually goes away! For the most part. I tested it with different values for the order parameter L's mobility, and evolution of the system changes quite a bit. When I set both the mobility and both diffusion coefficients to 1, I get the final state as
![image](https://user-images.githubusercontent.com/56096041/113355714-0812ec00-92f6-11eb-903a-5d03418205b0.png)
Which, as I understand it, is close to what we would expect for the chemical potential mu. The solver output is:
```
Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          1
  Spatial Dimension:       1
  Nodes:
    Total:                 101
    Local:                 14
  Elems:
    Total:                 100
    Local:                 13
  Num Subdomains:          1
  Num Partitions:          8
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                202
  Num Local DOFs:          28
  Variables:               { "eta" "mu" }
  Finite Element Types:    "LAGRANGE"
  Approximation Orders:    "FIRST"

Auxiliary System:
  Num DOFs:                1500
  Num Local DOFs:          195
  Variables:               { "local_GP_FE" "local_FE" "M_mu" "chi" "drho_eta" ... "rho" "rhoEq_l" "w_l"
                             "w_loc" "w_s" }
  Finite Element Types:    "MONOMIAL"
  Approximation Orders:    "CONSTANT"

Execution Information:
  Executioner:             Transient
  TimeStepper:             IterationAdaptiveDT
  Solver Mode:             Preconditioned JFNK
  MOOSE Preconditioner:    SMP


Time Step 0, time = 0

Postprocessor Values:
+----------------+----------------+----------------+
| time           | total_FE       | total_GP_FE    |
+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
+----------------+----------------+----------------+


Time Step 1, time = 1e-08, dt = 1e-08

Performing automatic scaling calculation

 0 Nonlinear |R| = 3.167606e-06
      0 Linear |R| = 3.167606e-06
      1 Linear |R| = 1.755020e-09
  Linear solve converged due to CONVERGED_RTOL iterations 1
 1 Nonlinear |R| = 1.755160e-09
Nonlinear solve converged due to CONVERGED_FNORM_ABS iterations 1
 Solve Converged!

Postprocessor Values:
+----------------+----------------+----------------+
| time           | total_FE       | total_GP_FE    |
+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e-08 |   2.117288e-01 |  -2.158171e-01 |
+----------------+----------------+----------------+
```

Just to give more information, if I keep diffusion equal to one and set L = 10, the system evolves to
![image](https://user-images.githubusercontent.com/56096041/113356092-9c7d4e80-92f6-11eb-995d-9bdb48bcbbf9.png)
And if L = 100, I get
![image](https://user-images.githubusercontent.com/56096041/113356299-df3f2680-92f6-11eb-85b2-f4ad4f7c77aa.png)
If I instead reduce L to 0.1, I get
![image](https://user-images.githubusercontent.com/56096041/113356486-24fbef00-92f7-11eb-89f6-479b497462b4.png)

The best of these is when L and the diffusivities are equal to 1, which almost feels like blind luck that I happened to find values that behave well. When I return the diffusivity to its original value, or to a constant value on the order to the liquid diffusivity, and then reduce L proportionally, I get the same problem as before where nothing evolves at all. I'm not sure how I can exploit the working model (where L=D=1) to choose a value for L that will work, there must be another piece I'm missing that plays a role in the mobilities' interplay.

**Reply by laagesen** [reply]
Hey Scott, the example in the repository I mentioned before is also based on Plapp's 2011 paper. So you could compare that to the model you've developed to see if there are any differences in the implementation.

I think that you may need to make some adjustments to your solver tolerances, time step, or problem scaling. Looking at your solver output, I notice that the original residual at the start of your first time step is `|R| = 3.167606e-06`. Let me explain what the residual means. When we set the governing equations up, we normally move all terms to the LHS of the equation so that the RHS of the equation is 0. The "residual" is what the numerical solver is finding for the RHS of the equation (which we would like to be zero or close to it). 

In your case, the solver executes one nonlinear iteration and exits with a residual `|R| = 1.755160e-09`. Normally we would like to see the residual decrease much more than that, like 8 or 9 orders of magnitude decrease. So I think the anomalous results you are getting is because the problem is not being solved to sufficiently high accuracy. Although 10^-9 seems like a small number, what we consider small in this context depends on how your problem is scaled and your time step size. Since you start out with a pretty small residual, it needs to decrease quite a lot more than 3 orders of magnitude. 

First thing to do is figure out why the solver is exiting. The solver will exit if either the absolute or relative tolerance criteria is met. The relative tolerance criteria is set by the parameter `nl_rel_tol` in your input file. For that to succeed, the final residual divided by original residual (at the beginning of the time step) must be < `nl_rel_tol`. This is how we normally like the solver to exit, and a value of 10^-8 is usually a good value to use. You may also exit if the absolute tolerance criterion is met. This is controlled by the parameter `nl_abs_tol`; if your residual is smaller than this the solver declares success. The problem with this is that what constitutes a small number varies on your time step size, problem scaling and other factors. My guess of what is happening is that you have a value of something like `nl_abs_tol = 2e-9` set such that your solver is exiting prematurely. So if that is the case I would set it to a much lower number or remove that parameter completely (in which case the default value of 1e-50 will be used).

Once you figure that out and make sure that your `nl_rel_tol` is set to a reasonable number, you should be able to increase your time step size significantly. That should increase the initial residual and then when the relative tolerance criterion is met the solve will proceed.

I think the reason you saw some evolution with L=D=1 and none at all for the smaller numbers that you were previously using is that with smaller L and D, the initial residual was probably below a set value of `nl_abs_tol`, so the solver declared success at the beginning of the problem and didn't actually do anything.
